### PR TITLE
Normalize protocol URIs

### PR DESCRIPTION
**Pull request by diehuxx**
Resolves part of #265. PR to normalize schemas will come next. Thanks @dcrousso for pointing me in the right direction!

The general approach is: Anywhere we call `MessageStore#put` or `MessageStore#query` normalize the protocol string in the indexes/filters respective.

### CODE REVIEW

File: src/interfaces/protocols/handlers/protocols-configure.ts [hunk 1]
```diff
@@ -56,7 +58,8 @@ export class ProtocolsConfigureHandler implements MethodHandler {
       const { author } = protocolsConfigure;
       const indexes = {
         author,
-        ... message.descriptor
+        ... message.descriptor,
```
Comment (dcrousso) [inline src/interfaces/protocols/handlers/protocols-configure.ts:61]: NIT: we probably could/should move this to be before `author` just in case it's inside `message.descriptor` so that it's not overridden by the `...`

File: src/interfaces/protocols/handlers/protocols-query.ts [hunk 2]
```diff
@@ -32,11 +32,14 @@ export class ProtocolsQueryHandler implements MethodHandler {
       return MessageReply.fromError(e, 401);
     }
 
-    const query = {
+    const query: Filter = {
       interface : DwnInterfaceName.Protocols,
       method    : DwnMethodName.Configure,
-      ...message.descriptor.filter
```
Comment (dcrousso) [inline src/interfaces/protocols/handlers/protocols-query.ts:38]: NIT: i recognize that there's no need for this anymore since the only value in `filter` is `protocol`, but some part of me feels like we should still leave this here just for future proofing sake 🤔

File: src/interfaces/records/handlers/records-write.ts [hunk 3]
```diff
@@ -120,6 +121,9 @@ export async function constructRecordsWriteIndexes(
     recordId  : message.recordId,
     entryId   : await RecordsWrite.getEntryId(recordsWrite.author, recordsWrite.message.descriptor)
   };
+  if (descriptor.protocol !== undefined) {
```
Comment (dcrousso) [inline src/interfaces/records/handlers/records-write.ts:124]: NIT: rather than check against the `descriptor.protocol`, should we check `indexes.protocol` since ultimately we really only care about normalizing the `protocol` value in the `IndexLevel`?  some part of me feels like it's odd to have it be added to `indexes` only to then check some other place before overriding it

File: src/interfaces/records/messages/records-query.ts [hunk 4]
```diff
@@ -92,6 +93,10 @@ export class RecordsQuery extends Message<RecordsQueryMessage> {
       (filterCopy as Filter).dateCreated = rangeFilter;
     }
 
+    if (filter.protocol !== undefined) {
```
Comment (dcrousso) [inline src/interfaces/records/messages/records-query.ts:96]: ditto

File: src/utils/url.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+export function normalizeProtocolUrl(url: string): string {
+  const fullUrl = url.includes('://') ? url : `http://${url}`;
```
Comment (dcrousso) [inline src/utils/url.ts:2]: what if the URL contains `://` in the path/query/etc.?  perhaps we should `/^[^:]+:\/\/./.test(url)` instead?

File: src/utils/url.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  const { hostname, pathname } = new URL(fullUrl);
```
Comment (dcrousso) [inline src/utils/url.ts:3]: why not include the `search` as well?  i feel like that's a pretty commonly used way of distinguishing resources on many webservers

NIT: we may want to `try { ... } catch { }` this just in case

File: tests/interfaces/protocols/handlers/protocols-query.spec.ts [hunk 7]
```diff
@@ -93,6 +100,88 @@ describe('ProtocolsQueryHandler.handle()', () => {
...
+      // Sort messages into lexicographic order. We only allow protocol overwrites if the new message CID
+      // has higher lexicographic value.
+      const equivalentProtocols = [protocol1, protocolTrailingSlash, protocolParams, protocolCapitalized];
+      let messageDataWithCid: (GenerateProtocolsConfigureOutput & { cid: string })[] = [];
+      for (const messageData of equivalentProtocols) {
+        const cid = await Message.getCid(messageData.message);
+        messageDataWithCid.push({ cid, ...messageData });
+      }
+
+      messageDataWithCid = messageDataWithCid.sort((messageDataA, messageDataB) => {
```
Comment (dcrousso) [inline tests/interfaces/protocols/handlers/protocols-query.spec.ts:141]: `Array.prototype.sort` is in-place, so no need to re-assign it :)

File: tests/interfaces/protocols/handlers/protocols-query.spec.ts [hunk 8]
```diff
@@ -93,6 +100,88 @@ describe('ProtocolsQueryHandler.handle()', () => {
...
+
+      // Configure protocols. Each one overwrites the previous
+      for (const protocol of messageDataWithCid) {
+        const configureReply = await dwn.processMessage(alice.did, protocol.message, protocol.dataStream);
+        expect(configureReply.status.code).to.equal(202);
+      }
+
+      // configure several more protocols, none of which match `example.com`
+      const protocolSlashFoo = await TestDataGenerator.generateProtocolsConfigure({ requester: alice, protocol: 'example.com/foo' });
+      const protocolSubdomainFoo = await TestDataGenerator.generateProtocolsConfigure({ requester: alice, protocol: 'foo.example.Com' });
```
Comment (dcrousso) [inline tests/interfaces/protocols/handlers/protocols-query.spec.ts:153]: is the `Com` a typo or intentional?  if the latter, can you explain why?

File: tests/interfaces/protocols/handlers/protocols-query.spec.ts [hunk 9]
```diff
@@ -93,6 +100,88 @@ describe('ProtocolsQueryHandler.handle()', () => {
...
+        protocolDefinition
+      });
+      const protocolCapitalized = await TestDataGenerator.generateProtocolsConfigure({
+        requester : alice,
+        protocol  : 'EXAMPLE.Com',
+        protocolDefinition
+      });
+
+      // Sort messages into lexicographic order. We only allow protocol overwrites if the new message CID
+      // has higher lexicographic value.
```
Comment (dcrousso) [inline tests/interfaces/protocols/handlers/protocols-query.spec.ts:133]: wow really?  that seems like an odd way of restricting things.

we dont sort based on the `dateModified` or anything like that?

File: tests/interfaces/protocols/handlers/protocols-query.spec.ts [hunk 10]
```diff
@@ -93,6 +100,88 @@ describe('ProtocolsQueryHandler.handle()', () => {
...
+      expect(reply.status.code).to.equal(200);
+      expect(reply.entries?.length).to.equal(1);
+
+      const resultProtocols = reply.entries!.map(entry =>
+        (entry.descriptor as ProtocolsConfigureDescriptor).protocol
+      );
+
+      // Expect equivalent protocol with highest lexicographic value to remain
+      expect(resultProtocols).to.contain(
+        messageDataWithCid[messageDataWithCid.length-1].message.descriptor.protocol
```
Comment (dcrousso) [inline tests/interfaces/protocols/handlers/protocols-query.spec.ts:177]: Style: i think there should be spaces around the `-`

File: tests/interfaces/records/handlers/records-query.spec.ts [hunk 11]
```diff
@@ -112,6 +112,65 @@ describe('RecordsQueryHandler.handle()', () => {
...
+        });
+        const writeReply = await dwn.processMessage(alice.did, write.message, write.dataStream);
+        expect(writeReply.status.code).to.equal(202);
+      }
+
+      // Query for matching protocol
+      const messageData = await TestDataGenerator.generateRecordsQuery({
+        requester : alice,
+        filter    : {
+          protocol: protocol1
```
Comment (dcrousso) [inline tests/interfaces/records/handlers/records-query.spec.ts:153]: should we also do a query for `protocol2` to make sure that it also matches the same entries?

**Review by dcrousso — APPROVED** [review]

File: tests/interfaces/protocols/handlers/protocols-query.spec.ts (hunk 9 above)
Comment (diehuxx) [reply tests/interfaces/protocols/handlers/protocols-query.spec.ts:133]: Yea I don't understanding why. @csuwildcat Why do we do this? I know I pinged in Slack, but asking here too so we can get an opened source'd explanation.

File: src/utils/url.ts (hunk 6 above)
Comment (diehuxx) [reply src/utils/url.ts:3]: Good call on try/catch. will add that today.

Excluding `search` because there is some conversation lately around how to configure different "roles" of a protocol. For example, for TBDex a DWN may want to set up as an "issuer" or as a "PFI" or a "wallet". Each of those roles would use the TBDex protocol a bit differently. The current idea that has been tossed around DWN office hours is to have that "role" be specified in a query param. So they should be normalized out.

This is a very very WIP idea, so we may end up changing it in the coming months (or even weeks), so if keeping `search` ends up being the right approach, I'm open to changing it later.

Comment (dcrousso) [reply src/utils/url.ts:3]: hmmm if that's the case then how would someone query for a specific "role" of the same protocol?

since we're always normalizing the `protocol` anywhere we send it to `MessageStore` (and therefore `IndexLevel`), would it be up to the caller to do their own filtering?

Comment (dcrousso) [reply src/utils/url.ts:3]: ah sorry i shouldve been clearer.  i was specifically referring to using `RecordsQuery` to something like

> get all messages with `{ protocol: 'example.com?role=issuer' }`

AFAICT what you currently have in this PR would return everything matching `{ protocol: 'example.com' }`, which could include other things (e.g. `{ protocol: 'example.com?role=other' }`) that the caller might not expect

we may want to make the normalization even more minimal than what's there now, such as only

- always having a `protocol` of `http://`
- removing trailing `'/'` from the `pathname`
- removing `port`
- removing `password`
- sorting the `search`
- removing `hash`

IMO this is part of why i dont think it's a good idea to do normalization, as we're effectively constraining the possibilities of URLs without any sort of specification as to what those constraints are (e.g. it's totally valid for a different port to return a different result).  i understand that we're trying to make the developer experience better by trying to avoid (possibly) common mistakes (e.g. adding a trailing `'/'`), but i think that could just result in developers being more confused by seemingly magical stuff happening instead of exactly what they asked for

---

as far as what you said, i think so long as `example.com?role=issuer` and `example.com?role=pfi` resolve to different things then that's fine :)

File: tests/interfaces/protocols/handlers/protocols-query.spec.ts (hunk 8 above)
Comment (diehuxx) [reply tests/interfaces/protocols/handlers/protocols-query.spec.ts]: typo. fixed

File: tests/interfaces/records/handlers/records-query.spec.ts (hunk 11 above)
Comment (diehuxx) [reply tests/interfaces/records/handlers/records-query.spec.ts:173]: good idea. added

File: src/utils/url.ts (hunk 6 above)
Comment (diehuxx) [reply src/utils/url.ts]: I'm starting to agree with you as I jam on edge cases. Lemme think about this a bit

File: tests/interfaces/protocols/handlers/protocols-query.spec.ts (hunk 9 above)
Comment (thehenrytsai) [reply tests/interfaces/protocols/handlers/protocols-query.spec.ts:133]: @dcrousso and @diehuxx, the comment is not entirely correct. We only resort to lexicographical order if all messages for the same ID are written with the exact same timestamp, which is the case here in the context of testing, but not a commonly expected scenario in the real-world.

**Review by thehenrytsai — COMMENTED** [review]
Seeing all the places that had `normalizeProtocolUrl()` inserted, I am wondering if it is better/worse to use prevention as a strategy to keep the internals clean? ie. do the normalization in `create()` methods and perform message validation upon request...

**Comment by diehuxx** [comment]
> do the normalization in create() methods and perform message validation upon request...

Not sure what you mean by perform message validation upon request?

File: tests/interfaces/protocols/handlers/protocols-query.spec.ts (hunk 9 above)
Comment (diehuxx) [reply tests/interfaces/protocols/handlers/protocols-query.spec.ts:133]: Ahh okay that makes more sense. Was very confused. Thanks!

File: src/interfaces/protocols/handlers/protocols-query.ts (hunk 2 above)
Comment (diehuxx) [reply src/interfaces/protocols/handlers/protocols-query.ts:38]: added back in
